File: Examples/StereoKitTest/Tools/LogWindow.cs
// SPDX-License-Identifier: MIT
// The authors below grant copyright rights under the MIT license:
// Copyright (c) 2023-2026 Nick Klingensmith
// Copyright (c) 2023-2026 Qualcomm Technologies, Inc.

using System;
using System.Collections.Generic;

namespace StereoKit.Framework
{
	public class LogWindow : IStepper
	{
		struct LogItem
		{
			public LogLevel level;
			public string   text;
			public int      count;
			public float    height; // Cached layout height (m), incl. padding.
		}
		// One entry per log line that intersects the viewport. Rebuilt only
		// when the scroll, content, or width changes - not every frame.
		struct VisibleLine
		{
			public int   index; // Index into 'items'.
			public float y;     // Entry top, viewport-relative (0 = viewport top, down is -).
		}
		private bool enabled = false;
		public  bool Enabled { get => enabled; set { enabled = value; pose = UI.PopupPose(); } }

		public Pose pose = Pose.Identity;

		TextStyle styleDiag;
		TextStyle styleInfo;
		TextStyle styleWarn;
		TextStyle styleErr;
		TextStyle styleGraph;

		// lineSpacing sets the distance between every line equally - both separate
		// entries and lines within a wrapped entry - so the two look identical.
		const float lineSpacing = 1.4f; // Line spacing as a fraction of font height (TextStyle.LineHeightPct).
		// Sizes in lines of log text, scaled to meters in Initialize so they track
		// the font size.
		const float viewLines  = 16;   // Viewport height, in lines.
		const float badgeLines = 1.5f; // Badge width, in lines.

		float rowPad     = 0; // Per-entry padding (m); derived from lineSpacing.
		float viewHeight = 0; // Viewport height (m); viewLines scaled in Initialize.
		float badgeSize  = 0; // Badge box width (m); badgeLines scaled in Initialize.

		List<LogItem>     items   = new List<LogItem>();
		List<VisibleLine> visible = new List<VisibleLine>();

		float scroll        = 0;     // Scroll offset (m) from the top (oldest) of the content.
		float scrollMax     = 0;     // Largest valid scroll, so the newest line sits at the bottom.
		float contentHeight = 0;     // Total height (m) of every log item.
		bool  stickBottom   = true;  // Keep the view pinned to the newest line.
		int   measuredCount = 0;     // How many items have a valid cached height.
		float measureWidth  = 0;     // Width the cached heights were measured against.
		float cachedScroll  = -1;    // Scroll value the visible list was built for.
		bool  viewDirty     = true;

		DiagGraph fpsGraph     = new DiagGraph(0, 0);
		DiagGraph cpuPerfGraph = new DiagGraph(0, 0);
		DiagGraph gpuPerfGraph = new DiagGraph(0, 0);

		public LogWindow()
		{
			Log.Subscribe(OnLog);
			Enabled = true;
		}

		public bool Initialize()
		{
			Font logFnt = Font.FromFamily("monospace");
			styleDiag  = TextStyle.FromFont(logFnt, 0.008f, Color.HSV(1, 0, 0.7f));
			styleInfo  = TextStyle.FromFont(logFnt, 0.008f, Color.HSV(1, 0, 1));
			styleWarn  = TextStyle.FromFont(logFnt, 0.008f, Color.HSV(0.17f, 0.7f, 1));
			styleErr   = TextStyle.FromFont(logFnt, 0.008f, Color.HSV(1, 0.7f, 1));
			styleGraph = TextStyle.FromFont(logFnt, 1,      Color.HSV(1, 0, 1));
			
			// Ensure the text is on the UI render layer
			styleDiag .RenderLayer = RenderLayer.UI;
			styleInfo .RenderLayer = RenderLayer.UI;
			styleWarn .RenderLayer = RenderLayer.UI;
			styleErr  .RenderLayer = RenderLayer.UI;
			styleGraph.RenderLayer = RenderLayer.UI;

			// Share the line spacing across every log style.
			styleDiag.LineHeightPct = lineSpacing;
			styleInfo.LineHeightPct = lineSpacing;
			styleWarn.LineHeightPct = lineSpacing;
			styleErr .LineHeightPct = lineSpacing;

			// One line baseline to baseline; rowPad tops a single entry up to a full
			// line (SizeLayout only returns the layout height), matching wrapped lines.
			float line = styleInfo.TotalHeight * lineSpacing;
			rowPad     = line - styleInfo.LayoutHeight;
			viewHeight = line * viewLines;
			badgeSize  = line * badgeLines;
			return true;
		}
		public void Shutdown()
			=> Log.Unsubscribe(OnLog);
		

		void OnLog(LogLevel level, string logText)
		{
			// Collapse a repeat of the previous line into a count on that entry.
			int last = items.Count - 1;
			if (last >= 0 && items[last].text == logText)
			{
				LogItem item = items[last];
				item.count++;
				items[last] = item;
				return;
			}
			// New items append at the bottom, so existing scroll offsets stay valid;
			// DrawLogs measures the new height and follows it if pinned to the bottom.
			items.Add(new LogItem { level = level, text = logText, count = 1 });
		}

		public void Step()
		{
			if (!Enabled) return;

			UI.WindowBegin("Log", ref pose, new Vec2(40, 0) * U.cm);

			DrawLogs();
			UI.HSeparator();
			DrawGraphs();

			UI.WindowEnd();
		}

		TextStyle StyleFor(LogLevel level) => level switch
		{
			LogLevel.Error      => styleErr,
			LogLevel.Warning    => styleWarn,
			LogLevel.Info       => styleInfo,
			LogLevel.Diagnostic => styleDiag,
			_                   => styleInfo,
		};

		// Fills 'visible' with each item that intersects the viewport, walking the
		// content from the oldest (top) to newest (bottom). Each line stores only
		// its top offset; clipping at both viewport edges is handled at draw time
		// (see DrawLogs), so partially visible lines at the top and bottom are kept.
		void RebuildVisible()
		{
			visible.Clear();
			float o = 0; // Content offset (m) of the current item's top.
			for (int i = 0; i < items.Count; i++)
			{
				float h    = items[i].height;
				float topY = scroll - o; // Item top, viewport-relative (0 = top, down is -).
				o += h;

				if (topY >= h)           continue; // Entirely above the viewport.
				if (topY <= -viewHeight) break;    // Entirely below the viewport.

				visible.Add(new VisibleLine { index = i, y = topY });
			}
		}

		void DrawLogs()
		{
			float width     = UI.LayoutRemaining.x;
			float sliderW   = UI.LineHeight*0.6f;
			float textWidth = width - sliderW; // Stop short of the slider.

			// Measure & cache the layout height of any item not yet measured (or all
			// of them if the width changed, since wrapping depends on it), keeping
			// contentHeight and scrollMax current. Cheap unless something changed.
			if (textWidth != measureWidth) { measuredCount = 0; measureWidth = textWidth; contentHeight = 0; }
			if (measuredCount != items.Count)
			{
				for (int i = measuredCount; i < items.Count; i++)
				{
					LogItem item = items[i];
					item.height   = Text.SizeLayout(item.text, StyleFor(item.level), textWidth).y + rowPad;
					items[i]      = item;
					contentHeight += item.height;
				}
				measuredCount = items.Count;
				scrollMax     = Math.Max(0, contentHeight - viewHeight);
				viewDirty     = true;
			}

			if (stickBottom) scroll = scrollMax;

			UI.LayoutPushCut(UICut.Top, viewHeight);
			UI.LayoutPushCut(UICut.Right, sliderW);
			// The slider handle tracks the content directly: bottom of the track
			// is the newest line, top is the oldest.
			if (UI.VSlider("scroll", ref scroll, 0, scrollMax))
				stickBottom = scroll >= scrollMax - 0.0001f;
			UI.LayoutPop();

			// Rebuild the cached visible list only when the scroll or content moved.
			scroll = Math.Clamp(scroll, 0, scrollMax);
			if (scroll != cachedScroll || viewDirty)
			{
				RebuildVisible();
				cachedScroll = scroll;
				viewDirty    = false;
			}

			Vec3 start = UI.LayoutAt;
			UI.LayoutReserve(V.XY(width, viewHeight));

			// Text.Add's 'size' is also its clip rectangle, and it's computed from
			// the pivot *before* offX/offY are applied - so the offset slides the text
			// independently of the clip box. We anchor the box to the whole viewport
			// and push each line into place with offY, which clips top *and* bottom in
			// one uniform top-anchored pass. The badge shares the text's plane so it
			// can't parallax-drift against it.
			float  z        = start.z - 0.004f;
			Matrix textTr   = Matrix.T(start.x,             start.y, z);
			Matrix badgeTr  = Matrix.T(start.x - textWidth, start.y, z);
			Vec2   textBox  = V.XY(textWidth, viewHeight);
			Vec2   badgeBox = V.XY(badgeSize, viewHeight);
			foreach (VisibleLine line in visible)
			{
				LogItem item = items[line.index];
				Text.Add(item.text, textTr, textBox, TextFit.Clip | TextFit.Wrap, StyleFor(item.level), Pivot.TopLeft, Align.TopLeft, 0, line.y);

				// Right-aligned duplicate-count badge, just inside the slider.
				if (item.count > 1)
					Text.Add(item.count.ToString(), badgeTr, badgeBox, TextFit.Clip, styleInfo, Pivot.TopRight, Align.TopRight, 0, line.y);
			}
			UI.LayoutPop();
		}

		void DrawGraphs()
		{
			float frameMs = 1000.0f / Device.DisplayRefreshRate;
			fpsGraph.min      = frameMs * 0.5f;
			fpsGraph.max      = frameMs * 1.5f;
			cpuPerfGraph.max  = frameMs * 1.5f;
			gpuPerfGraph.max  = frameMs * 1.5f;

			float thirdWidth = (UI.LayoutRemaining.x - UI.Settings.gutter*2) / 3.0f;
			Vec2  graphSize  = V.XY(thirdWidth, UI.LineHeight);

			void Graph(ref DiagGraph graph, string label, float ms)
			{
				graph.Add(ms);
				graph.Draw(styleGraph, label, UI.LayoutAt + V.XYZ(0,0,-0.004f), graphSize);
				UI.LayoutReserve(graphSize);
				UI.SameLine();
			}
			Graph(ref fpsGraph,     "Frame (ms)", Time.Stepf*1000);
			Graph(ref cpuPerfGraph, "CPU (ms)",   Time.PerfCPUus / 1000.0f);
			Graph(ref gpuPerfGraph, "GPU (ms)",   Time.PerfGPUus / 1000.0f);
		}

		struct DiagGraph
		{
			float[]     buffer;
			LinePoint[] points;
			int         curr;
			public float min;
			public float max;
			float       avg;

			public DiagGraph(float min, float max)
			{
				this.min = min;
				this.max = max;
				curr = 0;
				avg  = 0;

				buffer = new float[100];
				points = new LinePoint[100];
				for (int i = 0; i < points.Length; i++)
				{
					points[i].color     = new Color32(255,255,255,100);
					points[i].thickness = 0.001f;
				}
			}

			public void Add(float value)
			{
				buffer[curr] = value;
				curr = (curr + 1) % buffer.Length;
				avg = avg + (value - avg) * 0.1f;
			}

			public void Draw(TextStyle style, string label, Vec3 at, Vec2 size)
			{
				const float labelWidth = 0.01f;
				float       textScale  = size.y * 0.15f;
				size.x -= labelWidth;

				float step = 1f/(points.Length-1);
				for (int i = 0; i < points.Length; i++)
				{
					float pct = i * step;
					int   idx = (curr + i) % points.Length;
					float val = Math.Clamp((buffer[idx] - min) / (max - min), 0.0f, 1.0f);
					points[i].pt = new Vec3(at.x - pct*size.x, at.y-size.y + val*size.y, at.z);
				}
				Lines.Add(points);

				string display = $"{label} {avg:F3}";
				Text.Add(((int)max).ToString(), Matrix.TS(at.x-size.x, at.y,        at.z, textScale), V.XY(labelWidth*textScale, 1), TextFit.Overflow, style, Pivot.TopLeft,    Align.TopLeft);
				Text.Add(((int)min).ToString(), Matrix.TS(at.x-size.x, at.y-size.y, at.z, textScale), V.XY(labelWidth*textScale, 1), TextFit.Overflow, style, Pivot.BottomLeft, Align.BottomLeft);
				Text.Add(display,               Matrix.TS(at.x,        at.y,        at.z, textScale),                                                  style, Pivot.TopLeft,    Align.TopLeft);
				Lines.Add(at+V.XY0(-Text.SizeLayout(display, style).x*textScale - 0.002f, 0), at + V.XY0(-size.x, 0), new Color32(255,255,255,20), 0.001f);
				Lines.Add(at+V.XY0(0, -size.y), at + V.XY0(-size.x, -size.y), new Color32(255, 255, 255, 20), 0.001f);
			}
		}
	}
}

File: Examples/StereoKitCTest/log_window.cpp
#include "log_window.h"

#include <stereokit.h>
#include <stereokit_ui.h>

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

using namespace sk;

///////////////////////////////////////////

struct log_item_t {
	log_  level;
	char* text;
	int   count;
	float height; // Cached layout height (m), incl. padding.
};
// One entry per log line that overlaps the viewport. Rebuilt only when the
// scroll, content, or width changes - not every frame.
struct visible_line_t {
	int   index; // Index into 'items'.
	float y;     // Entry top, viewport-relative (0 = viewport top, down is -).
};

// A small ring-buffer line graph for per-frame timing values.
struct diag_graph_t {
	float        buffer[100];
	line_point_t points[100];
	int          curr;
	float        min;
	float        max;
	float        avg;
};

static void diag_graph_init(diag_graph_t* graph, float min, float max);
static void diag_graph_add (diag_graph_t* graph, float value);
static void diag_graph_draw(diag_graph_t* graph, text_style_t style, const char* label, vec3 at, vec2 size);

///////////////////////////////////////////

// One spacing value for both the gaps between entries and the lines within a
// wrapped entry, so the two look identical.
static const float line_spacing = 1.4f; // Fraction of font height.
// Sizes in lines of log text, scaled to meters in ensure_styles so they track
// the font size.
static const float view_lines  = 16;   // Viewport height, in lines.
static const float badge_lines = 1.5f; // Badge width, in lines.

static bool styles_ready = false; // Styles are built lazily, after StereoKit is up.

static text_style_t style_diag;
static text_style_t style_info;
static text_style_t style_warn;
static text_style_t style_err;
static text_style_t style_graph;

static float row_pad     = 0; // Per-entry padding (m); derived from line_spacing.
static float view_height = 0; // Viewport height (m); view_lines scaled in init.
static float badge_size  = 0; // Badge box width (m); badge_lines scaled in init.

static log_item_t* items       = nullptr;
static int         items_count = 0;
static int         items_cap   = 0;

static visible_line_t* visible       = nullptr;
static int             visible_count = 0;
static int             visible_cap   = 0;

static float scroll         = 0;    // Scroll offset (m) from the top (oldest) of the content.
static float scroll_max     = 0;    // Largest valid scroll, so the newest line sits at the bottom.
static float content_height = 0;    // Total height (m) of every log item.
static bool  stick_bottom   = true; // Keep the view pinned to the newest line.
static int   measured_count = 0;    // How many items have a valid cached height.
static float measure_width  = 0;    // Width the cached heights were measured against.
static float cached_scroll  = -1;   // Scroll value the visible list was built for.
static bool  view_dirty     = true;

static diag_graph_t fps_graph;
static diag_graph_t cpu_perf_graph;
static diag_graph_t gpu_perf_graph;

///////////////////////////////////////////

static float clampf(float v, float min, float max) {
	return v < min ? min : (v > max ? max : v);
}

// Grows a heap array to fit one more element, doubling capacity. Returns the
// (possibly relocated) data pointer.
static void* arr_ensure(void* data, int count, int* cap, size_t stride) {
	if (count < *cap) return data;
	*cap = *cap < 16 ? 16 : *cap * 2;
	return realloc(data, (size_t)*cap * stride);
}

static char* str_copy(const char* src) {
	size_t size = strlen(src) + 1;
	char*  dst  = (char*)malloc(size);
	memcpy(dst, src, size);
	return dst;
}

static text_style_t style_for(log_ level) {
	switch (level) {
	case log_error:      return style_err;
	case log_warning:    return style_warn;
	case log_inform:     return style_info;
	case log_diagnostic: return style_diag;
	default:             return style_info;
	}
}

///////////////////////////////////////////

static void on_log_window(void*, log_ level, const char* log_c_str) {
	// Collapse a repeat of the previous line into a count on that entry.
	if (items_count > 0 && strcmp(items[items_count-1].text, log_c_str) == 0) {
		items[items_count-1].count++;
		return;
	}
	// Append at the bottom so existing scroll offsets stay valid; draw_logs
	// measures the new height and follows it when pinned.
	items = (log_item_t*)arr_ensure(items, items_count, &items_cap, sizeof(log_item_t));
	items[items_count++] = log_item_t{ level, str_copy(log_c_str), 1, 0 };
}

///////////////////////////////////////////

void log_window_init() {
	log_subscribe(on_log_window);
}

///////////////////////////////////////////

// Build the text styles and derived metrics the first time we draw. Kept out of
// log_window_init so that init can run before sk_init to catch startup logs.
static void ensure_styles() {
	if (styles_ready) return;
	styles_ready = true;

	font_t log_fnt = font_create_family("monospace");
	style_diag  = text_make_style(log_fnt, 0.008f, color_hsv(1,    0,    0.7f, 1));
	style_info  = text_make_style(log_fnt, 0.008f, color_hsv(1,    0,    1,    1));
	style_warn  = text_make_style(log_fnt, 0.008f, color_hsv(0.17f,0.7f, 1,    1));
	style_err   = text_make_style(log_fnt, 0.008f, color_hsv(1,    0.7f, 1,    1));
	style_graph = text_make_style(log_fnt, 1,      color_hsv(1,    0,    1,    1));
	font_release(log_fnt);

	text_style_set_render_layer(style_diag,  render_layer_ui);
	text_style_set_render_layer(style_info,  render_layer_ui);
	text_style_set_render_layer(style_warn,  render_layer_ui);
	text_style_set_render_layer(style_err,   render_layer_ui);
	text_style_set_render_layer(style_graph, render_layer_ui);

	text_style_set_line_height_pct(style_diag, line_spacing);
	text_style_set_line_height_pct(style_info, line_spacing);
	text_style_set_line_height_pct(style_warn, line_spacing);
	text_style_set_line_height_pct(style_err,  line_spacing);

	// row_pad tops a single entry's layout height up to a full line, matching the
	// baseline-to-baseline spacing of wrapped lines.
	float line  = text_style_get_total_height(style_info) * line_spacing;
	row_pad     = line - text_style_get_layout_height(style_info);
	view_height = line * view_lines;
	badge_size  = line * badge_lines;

	diag_graph_init(&fps_graph,      0, 0);
	diag_graph_init(&cpu_perf_graph, 0, 0);
	diag_graph_init(&gpu_perf_graph, 0, 0);
}

///////////////////////////////////////////

void log_window_shutdown() {
	log_unsubscribe(on_log_window);

	for (int i = 0; i < items_count; i++)
		free(items[i].text);
	free(items);
	free(visible);

	items          = nullptr; items_count   = 0; items_cap   = 0;
	visible        = nullptr; visible_count = 0; visible_cap = 0;
	measured_count = 0;
	content_height = 0;
}

///////////////////////////////////////////

// Fills 'visible' with every item that overlaps the viewport, walking top
// (oldest) to bottom. Edge clipping happens at draw time, so partial lines are
// kept.
static void rebuild_visible() {
	visible_count = 0;
	float o = 0; // Content offset (m) of the current item's top.
	for (int i = 0; i < items_count; i++) {
		float h     = items[i].height;
		float top_y = scroll - o; // Item top, viewport-relative (0 = top, down is -).
		o += h;

		if (top_y >= h)            continue; // Entirely above the viewport.
		if (top_y <= -view_height) break;    // Entirely below the viewport.

		visible = (visible_line_t*)arr_ensure(visible, visible_count, &visible_cap, sizeof(visible_line_t));
		visible[visible_count++] = visible_line_t{ i, top_y };
	}
}

///////////////////////////////////////////

static void draw_logs() {
	float width      = ui_layout_remaining().x;
	float slider_w   = ui_line_height()*0.6f;
	float text_width = width - slider_w; // Stop short of the slider.

	// Measure & cache layout height for any new items (or all of them when the
	// width changed, since wrapping depends on it). Cheap when nothing changed.
	if (text_width != measure_width) { measured_count = 0; measure_width = text_width; content_height = 0; }
	if (measured_count != items_count) {
		for (int i = measured_count; i < items_count; i++) {
			items[i].height = text_size_layout_constrained(items[i].text, style_for(items[i].level), text_width).y + row_pad;
			content_height += items[i].height;
		}
		measured_count = items_count;
		scroll_max     = fmaxf(0, content_height - view_height);
		view_dirty     = true;
	}

	if (stick_bottom) scroll = scroll_max;

	ui_layout_push_cut(ui_cut_top,   view_height);
	ui_layout_push_cut(ui_cut_right, slider_w);
	// The slider handle tracks the content directly: bottom of the track is the
	// newest line, top is the oldest.
	if (ui_vslider("scroll", scroll, 0, scroll_max))
		stick_bottom = scroll >= scroll_max - 0.0001f;
	ui_layout_pop();

	scroll = clampf(scroll, 0, scroll_max);
	if (scroll != cached_scroll || view_dirty) {
		rebuild_visible();
		cached_scroll = scroll;
		view_dirty    = false;
	}

	vec3 start = ui_layout_at();
	ui_layout_reserve(vec2{ width, view_height });

	// text_add_in's 'size' doubles as the clip box and is taken at the pivot
	// before off_x/off_y, so off_y slides each line within a fixed top-anchored
	// clip - clipping top and bottom in one pass.
	float  z         = start.z - 0.004f;
	matrix text_tr   = matrix_t(vec3{ start.x,              start.y, z });
	matrix badge_tr  = matrix_t(vec3{ start.x - text_width, start.y, z });
	vec2   text_box  = vec2{ text_width, view_height };
	vec2   badge_box = vec2{ badge_size, view_height };
	for (int i = 0; i < visible_count; i++) {
		const visible_line_t& line = visible[i];
		const log_item_t&     item = items[line.index];
		text_add_in(item.text, text_tr, text_box, (text_fit_)(text_fit_clip | text_fit_wrap), style_for(item.level), pivot_top_left, align_top_left, 0, line.y);

		// Right-aligned duplicate-count badge, just inside the slider.
		if (item.count > 1) {
			char count_str[16];
			snprintf(count_str, sizeof(count_str), "%d", item.count);
			text_add_in(count_str, badge_tr, badge_box, text_fit_clip, style_info, pivot_top_right, align_top_right, 0, line.y);
		}
	}
	ui_layout_pop();
}

///////////////////////////////////////////

static void draw_graphs() {
	float frame_ms = 1000.0f / device_display_get_refresh_rate();
	fps_graph     .min = frame_ms * 0.5f;
	fps_graph     .max = frame_ms * 1.5f;
	cpu_perf_graph.max = frame_ms * 1.5f;
	gpu_perf_graph.max = frame_ms * 1.5f;

	float third_width = (ui_layout_remaining().x - ui_get_settings().gutter*2) / 3.0f;
	vec2  graph_size  = vec2{ third_width, ui_line_height() };

	auto graph = [&](diag_graph_t &g, const char* label, float ms) {
		diag_graph_add (&g, ms);
		diag_graph_draw(&g, style_graph, label, ui_layout_at() + vec3{0,0,-0.004f}, graph_size);
		ui_layout_reserve(graph_size);
		ui_sameline();
	};
	graph(fps_graph,      "Frame (ms)", time_stepf()*1000);
	graph(cpu_perf_graph, "CPU (ms)",   time_perf_cpu_us() / 1000.0f);
	graph(gpu_perf_graph, "GPU (ms)",   time_perf_gpu_us() / 1000.0f);
}

///////////////////////////////////////////

void log_window_update(pose_t* pose) {
	ensure_styles();

	ui_window_begin("Log", pose, vec2{ 40*cm2m, 0 });

	draw_logs();
	ui_hseparator();
	draw_graphs();

	ui_window_end();
}

///////////////////////////////////////////

static void diag_graph_init(diag_graph_t* graph, float min, float max) {
	graph->min  = min;
	graph->max  = max;
	graph->curr = 0;
	graph->avg  = 0;
	for (int i = 0; i < (int)(sizeof(graph->points)/sizeof(graph->points[0])); i++) {
		graph->buffer[i]           = 0;
		graph->points[i].color     = color32{255,255,255,100};
		graph->points[i].thickness = 0.001f;
	}
}

///////////////////////////////////////////

static void diag_graph_add(diag_graph_t* graph, float value) {
	int count = (int)(sizeof(graph->buffer)/sizeof(graph->buffer[0]));
	graph->buffer[graph->curr] = value;
	graph->curr = (graph->curr + 1) % count;
	graph->avg  = graph->avg + (value - graph->avg) * 0.1f;
}

///////////////////////////////////////////

static void diag_graph_draw(diag_graph_t* graph, text_style_t style, const char* label, vec3 at, vec2 size) {
	const float label_width = 0.01f;
	float       text_scale  = size.y * 0.15f;
	int         count       = (int)(sizeof(graph->points)/sizeof(graph->points[0]));
	size.x -= label_width;

	float step = 1.0f/(count-1);
	for (int i = 0; i < count; i++) {
		float pct = i * step;
		int   idx = (graph->curr + i) % count;
		float val = clampf((graph->buffer[idx] - graph->min) / (graph->max - graph->min), 0.0f, 1.0f);
		graph->points[i].pt = vec3{ at.x - pct*size.x, at.y-size.y + val*size.y, at.z };
	}
	line_add_listv(graph->points, count);

	char display[128];
	char num[16];
	snprintf(display, sizeof(display), "%s %.3f", label, graph->avg);
	snprintf(num,     sizeof(num),     "%d", (int)graph->max);
	text_add_in(num,     matrix_ts(vec3{at.x-size.x, at.y,        at.z}, vec3{text_scale,text_scale,text_scale}), vec2{label_width*text_scale, 1}, text_fit_overflow, style, pivot_top_left,    align_top_left);
	snprintf(num,     sizeof(num),     "%d", (int)graph->min);
	text_add_in(num,     matrix_ts(vec3{at.x-size.x, at.y-size.y, at.z}, vec3{text_scale,text_scale,text_scale}), vec2{label_width*text_scale, 1}, text_fit_overflow, style, pivot_bottom_left, align_bottom_left);
	text_add_at(display, matrix_ts(vec3{at.x,        at.y,        at.z}, vec3{text_scale,text_scale,text_scale}),                                                    style, pivot_top_left,    align_top_left);

	float display_w = text_size_layout(display, style).x;
	line_add(at + vec3{-display_w*text_scale - 0.002f, 0,       0}, at + vec3{-size.x, 0,       0}, color32{255,255,255,20}, color32{255,255,255,20}, 0.001f);
	line_add(at + vec3{0,                              -size.y, 0}, at + vec3{-size.x, -size.y, 0}, color32{255,255,255,20}, color32{255,255,255,20}, 0.001f);
}
